helm/config.go

```diff
@@ -0,0 +1,377 @@
+func Condition(condition string) func(*sharedFields) {
+	return func(shared *sharedFields) { shared.condition = condition }
+}
+
+// Set applies NodeModifier functions to the embedded sharedFields struct.
+func (shared *sharedFields) Set(modifiers ...NodeModifier) {
+	for _, modifier := range modifiers {
+		modifier(shared)
+	}
+}
+
+// Comment is a shared getter function for all node types.
+func (shared sharedFields) Comment() string {
+	return shared.comment
+}
+
+// Condition is a shared getter function for all node types.
+func (shared sharedFields) Condition() string {
+	return shared.condition
+}
+
+// Node is the interface implemented by all config node types.
+type Node interface {
+	// Every node will embed a sharedFields struct and inherit these methods:
+	Comment() string
+	Condition() string
+	Set(...NodeModifier)
+
+	// The write() method is specific to each node type. prefix will be the
+	// indented label, either "name:" or "-", depending on whether the node
+	// is part of an object or a list.
+	write(enc *Encoder, prefix string)
+}
+
+// Scalar represents a scalar value inside a list or object.
+type Scalar struct {
+	sharedFields
+	value string
+}
+
+// NewScalar creates a scalar node and initializes shared fields.
+func NewScalar(value string, modifiers ...NodeModifier) *Scalar {
+	scalar := &Scalar{value: value}
+	scalar.Set(modifiers...)
+	return scalar
+}
+
+func (scalar Scalar) write(enc *Encoder, prefix string) {
+	if strings.ContainsAny(scalar.value, "\n") {
+		// Scalars including newlines will be written using the "literal" YAML format.
+		fmt.Fprintln(enc.writer, prefix+" |-")
+		// Calculate proper indentation for data lines.
+		if strings.HasSuffix(prefix, ":") {
+			prefix = strings.Repeat(" ", strings.LastIndex(prefix, " ")+1+enc.indent)
+		} else {
+			prefix = strings.Repeat(" ", len(prefix)+1)
+		}
+		for _, line := range strings.Split(scalar.value, "\n") {
+			fmt.Fprintln(enc.writer, prefix+line)
+		}
+	} else {
+		fmt.Fprintln(enc.writer, prefix+" "+scalar.value)
+	}
+}
+
+// List represents an ordered list of unnamed nodes.
+type List struct {
+	sharedFields
+	nodes []Node
+}
+
+// NewList creates an empty list node and initializes shared fields.
+func NewList(modifiers ...NodeModifier) *List {
+	list := &List{}
+	list.Set(modifiers...)
+	return list
+}
+
+// Add one or more nodes at the end of the list.
+func (list *List) Add(nodes ...Node) {
+	list.nodes = append(list.nodes, nodes...)
+}
+
+func (list List) write(enc *Encoder, prefix string) {
+	for _, node := range list.nodes {
+		enc.writeNode(node, &prefix, strings.Repeat(" ", enc.indent-2)+"-")
+	}
+}
+
+// Objects store nodes in a list of name/node pairs instead of using a map so that
+// the nodes will be encoded in the same order in which they were added to the object.
+type namedNode struct {
+	name string
+	node Node
+}
+
+// Object represents an ordered lst of named nodes.
+type Object struct {
```

Comment:
I don't know; I took `Object` from the JSON spec. I thought about `Map` and `Hash` too, but don't really care for any of them either. I don't find `Mapping` any better as well, just different. :)

Note that the name works better once it is prefixed with the package name: `helm.Object`, `helm.List`, and `helm.Scalar`.

Reply:
JSON / JS is magical because Object _is_ their key-value type, as well as generic things with methods attached.  As far as I know YAML doesn't support methods on their hashes yet (thankfully).  FWIW, YAML does call them [mappings](http://www.yaml.org/spec/1.2/spec.html#id2790832). But then it also calls lists "sequences".

Reply:
Yeah, but YAML also calls lists "sequences" and I really don't want to use that term. 😄 

I would switch to `Map` if not for the fact that `map` is a reserved word. ☹️

Other names I rejected were `Array` and `Struct` because they feel like having a fixed capacity to me.

Any other ideas?